vktrace/vktrace_layer/vktrace_lib_trim_statetracker.cpp

```diff
@@ -0,0 +1,1772 @@
+            createdDescriptorSetLayouts.begin()->first));
+    }
+
+    while (createdDescriptorSets.size() != 0) {
+        remove_DescriptorSet(
+            static_cast<VkDescriptorSet>(createdDescriptorSets.begin()->first));
+    }
+
+    for (auto iter = m_cmdBufferPackets.begin();
+         iter != m_cmdBufferPackets.end(); ++iter) {
+        std::list<vktrace_trace_packet_header *> &packets = iter->second;
+        for (auto packetIter = iter->second.begin();
+             packetIter != iter->second.end(); ++packetIter) {
+            vktrace_trace_packet_header *pHeader = *packetIter;
+            vktrace_delete_trace_packet(&pHeader);
+        }
+        packets.clear();
+    }
+    m_cmdBufferPackets.clear();
+
+    for (auto packet = m_image_calls.begin(); packet != m_image_calls.end();
+         ++packet) {
+        vktrace_trace_packet_header *pHeader = *packet;
+        vktrace_delete_trace_packet(&pHeader);
+    }
+    m_image_calls.clear();
+
+    for (auto renderPassIter = m_renderPassVersions.begin();
+         renderPassIter != m_renderPassVersions.end(); ++renderPassIter) {
+        std::vector<VkRenderPassCreateInfo *> versions = renderPassIter->second;
+
+        for (uint32_t i = 0; i < versions.size(); i++) {
+            VkRenderPassCreateInfo *pCreateInfo = versions[i];
+
+            VkSubpassDescription *pSubPasses =
+                const_cast<VkSubpassDescription *>(pCreateInfo->pSubpasses);
+            for (uint32_t i = 0; i < pCreateInfo->subpassCount; i++) {
+                if (pCreateInfo->pSubpasses[i].pInputAttachments != nullptr) {
+                    delete[] pCreateInfo->pSubpasses[i].pInputAttachments;
+                }
+
+                if (pCreateInfo->pSubpasses[i].pColorAttachments != nullptr) {
+                    delete[] pCreateInfo->pSubpasses[i].pColorAttachments;
+                }
+
+                if (pCreateInfo->pSubpasses[i].pResolveAttachments != nullptr) {
+                    delete[] pCreateInfo->pSubpasses[i].pResolveAttachments;
+                }
+
+                if (pCreateInfo->pSubpasses[i].pDepthStencilAttachment !=
+                    nullptr) {
+                    delete[] pCreateInfo->pSubpasses[i].pDepthStencilAttachment;
+                }
+
+                if (pCreateInfo->pSubpasses[i].preserveAttachmentCount > 0) {
+                    delete[] pCreateInfo->pSubpasses[i].pPreserveAttachments;
+                }
+            }
+
+            if (pCreateInfo->pAttachments != nullptr) {
+                delete[] pCreateInfo->pAttachments;
+            }
+
+            delete pCreateInfo;
+        }
+        versions.clear();
+    }
+    m_renderPassVersions.clear();
+}
+
+//-------------------------------------------------------------------------
+void StateTracker::add_RenderPassCreateInfo(
+    VkRenderPass renderPass, const VkRenderPassCreateInfo *pCreateInfo) {
+    VkRenderPassCreateInfo *pCopyCreateInfo = new VkRenderPassCreateInfo();
```

Comment:
Would it be safe to add NULL checks to this function all over since multiple allocations involved?